Build an upload manager component to handle uploads

mbenadda opened the pull request:
## Purpose

We're building a new Marsha site whose main purpose (for now) is enabling users to easily upload multiple videos quickly to a specialized interface.

As part of this effort, we had to move away from the way uploads are handled in the LTI app as the requirements are different. However, we still wanted to mutualize the code that handles them so we can reuse components and have both apps' UI converge in the future.

## Proposal

We built an `<UploadManager />` component with an associated hook to manage everything related to uploads and provide complete information about the state of current uploads to any component(s) that might need it.

- [x] build the `<UploadManager />` and its related components
- [x] use it to replace the upload-related code in the LTI app
- [x] rework existing components to rely on it
- [x] remove the fictional `UPLOADING` state on uploadable objects

lunika commented on src/frontend/components/UploadManager/index.spec.tsx:
```diff
@@ -0,0 +1,248 @@
+      </UploadManager>,
+    );
+
+    {
```
Adding blocks like this really ease reading the code :+1:

lunika commented on src/frontend/components/ObjectStatusPicker/index.spec.tsx:
```diff
@@ -13,118 +18,367 @@ const {
+                  }),
+                  objectId: object.id,
+                  objectType: modelName.VIDEOS,
+                  progress: 60,
```
```suggestion
                  progress: 100,
```

To reflect the upload state.

lunika commented on src/frontend/components/ObjectStatusPicker/index.spec.tsx:
```diff
@@ -13,118 +18,367 @@ const {
-    screen.getByText('Error ❌');
-  });
+    it('renders status info for an object undergoing PROCESSING', () => {
+      // Here the object is not an upload that has just finished, but we for the PROCESSING state from the server.
```
I think words are missing at the end of this sentence.

lunika commented on src/frontend/components/UploadField/index.tsx:
```diff
@@ -2,28 +2,28 @@ import React, { useState } from 'react';
+
+export const UploadField = ({ objectType, objectId }: UploadFieldProps) => {
+  const { addUpload } = useUploadManager();
   const [file, setFile] = useState(undefined as Maybe<File>);
```
```suggestion
const [file, setFile] = useState<Maybe<File>>(undefined);
```

lunika reviewed (commented):
There is a regression when the video is uploaded and processing. We can again choose between uploading a video or configure a live. Configuring a live should not be possible

![Screenshot_2021-03-19 Screenshot](https://user-images.githubusercontent.com/767834/111762345-37ab0880-88a1-11eb-9af7-efba72ea3ef3.png)

lunika commented on src/frontend/components/ObjectStatusPicker/index.tsx:
```diff
@@ -88,39 +88,71 @@ const messages = defineMessages({
 });
 
+const isPossibleLiveObject = (object: UploadableObject): object is Video =>
+  !!(object as any).live_state;
```
No this is wrong, `live_state` is always in a video object.

jbpenrath replied:
Oh yes, I misread the video type.

mbenadda commented on src/frontend/components/ObjectStatusPicker/index.tsx:
```diff
@@ -88,39 +88,71 @@ const messages = defineMessages({
+export const ObjectStatusPicker = ({ object }: ObjectStatusPickerProps) => {
+  const { uploadManagerState } = useUploadManager();
+
+  if (isPossibleLiveObject(object) && object.live_state) {
```
`isPossibleLiveObject` is a type guard. As this is effectively a type check, TypeScript requires us to either force a typecast on `object` or use a type guard so the compiler knows what kind of object we have.

jbpenrath replied:
Yes yes but something is wrong in your type guard. In fact in Video, live_state is Nullable, but the check executed in `isPossibleLiveObject` return false if live_state is null.
So if you want a type guard to ensute that you have a video you should write this:
```
(object as any)?.live_state !== undefined
```

But currently you are testing twice that `live_state` is defined.

mbenadda replied:
I understand, there's no user-facing bug but the type check is incorrect and could create a bug later on. I'll change the test to what you suggested, which is indeed the correct one.

mbenadda commented:
> There is a regression when the video is uploaded and processing. We can again choose between uploading a video or configure a live. Configuring a live should not be possible and replacing it was not possible before. I don't know if it should be possible now.

Good catch, I updated the test for this component and changed the behavior in the component. My goal with this PR is to keep the same behavior 100% and just update the inner workings.

lunika commented on src/frontend/components/DashboardVideoPane/index.tsx:
```diff
@@ -143,7 +155,7 @@ export const DashboardVideoPane = ({ video }: DashboardVideoPaneProps) => {
 
   useEffect(() => {
-    if ([PENDING, UPLOADING, PROCESSING].includes(video.upload_state)) {
+    if ([PENDING, PROCESSING].includes(video.upload_state)) {
```
I think the `PENDING` test is useless. We now the exact state with the `uploadManagerState`. We can start polling once the upload in sucess and also when the video state is in `PROCESSING`. WDYT ?

lunika replied:
yes :+1:
